src/odemis/driver/rigol.py

```diff
@@ -0,0 +1,444 @@
+        except socket.error:
+            raise model.HwError("Failed to connect to '%s:%d', check that the Rigol "
+                                "Clock Generator is connected to the network, turned "
+                                "on, and correctly configured." % (host, port))
+
+        self.socket.settimeout(10) # s
+
+        # to acquire before sending anything on the socket
+        self._net_access = threading.Lock()
+        
+    def __del__(self):
+        self.terminate()
+        
+    def isConnected(self):
+        return self._is_connected
+
+    def terminate(self):
+        if self.remote_device:
+            self.remote_device.terminate()
+        if self._is_connected:
+            self.socket.close()
+        self._is_connected = False
+
+    def sendOrderCommand(self, cmd, val=""):
+        """
+        Sends one command, and don't expect any reply
+        cmd (str): command to send
+        val (str): value to send (if any)
+        raises:
+            IOError: if problem with sending/receiving data over the connection
+        """
+        if not self._is_connected:
+            raise IOError("Device %s not connected." % (self._host))
+
+        msg = "%s %s\n" % (cmd, val)
+
+        with self._net_access:
+            self.socket.sendall(msg)
+
+    def sendQueryCommand(self, cmd):
+        """
+        Sends one command, and don't expect any reply
+        cmd (str): command to send, without ?
+
+        raises:
+            IOError: if problem with sending/receiving data over the connection
+            NewFocusError: if error happened
+        """
+        if not self._is_connected:
+            raise IOError("Device %s not connected." % (self._host))
+        
+        msg = "%s\n" % cmd
+
+        with self._net_access:
+            self.socket.sendall(msg)
+
+            data = None
+            try:
+                data = self.socket.recv(4096)
+            except socket.timeout:
+                raise IOError("Device %s timed out. %s" %
+                              (self._host, msg.encode('string_escape')))
+
+            if not data:
+                time.sleep(0.01)
+                logging.debug("Received empty message")
+
+        return data.strip()
+
+
+class FakeDG1000Z(object):
+    '''
+    A simulated Rigol DG1000Z Clock Generator. 
+    Runs a listening thread that acts as a simulated device on localhost 
+    '''
+    def __init__(self):
+        
+        # parameters
+        self._error_state = False
+        self._output_buffer = ''
+        self.name = "SimWG"
+        logging.debug('%s: Starting simulated device' %  (self.name))
+        self._frequency = 1000
+        
+        # Set up listening socket
+        self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self.host = ('localhost', DEFAULT_PORT)
+        self.socket.bind(self.host)
+        
+        # Create a thread to listen for incoming commands
+        self._shutdown_flag = threading.Event()
```

Comment:
Is that thread safe? I think the shutdown flag is safer.

Reply:
In Python it's safe to change any variable concurrently. It's just not guaranteed which value will actually be used. 

In that very case, it would be safe, because the only possibility is to have `_shutdown_flag` change from `True` to `False`. Even if it was modified simultaneously in separate threads, eventually, it would be `False`. You can leave it to an Event, if you prefer.